pkg/providers/vsphere/virtualmachine/affinity.go

```diff
@@ -0,0 +1,260 @@
+
+	// Process required anti-affinity terms associated with zone topology.
+	requiredZoneLabels := extractTermLabelsByTopology(
+		vmCtx,
+		antiAffinity.RequiredDuringSchedulingPreferredDuringExecution,
+		corev1.LabelTopologyZone,
+	)
+	if len(requiredZoneLabels) > 0 {
+		placementPols = append(placementPols, &vimtypes.VmToVmGroupsAntiAffinity{
+			AntiAffinedVmGroupTags: requiredZoneLabels,
+			PolicyStrictness:       string(vimtypes.VmPlacementPolicyVmPlacementPolicyStrictnessRequiredDuringPlacementPreferredDuringExecution),
+			PolicyTopology:         string(vimtypes.VmPlacementPolicyVmPlacementPolicyTopologyVSphereZone),
+		})
+	}
+
+	// Process required anti-affinity terms associated with host topology.
+	requiredHostLabels := extractTermLabelsByTopology(
+		vmCtx,
+		antiAffinity.RequiredDuringSchedulingPreferredDuringExecution,
+		corev1.LabelHostname,
+	)
+	if len(requiredHostLabels) > 0 {
+		placementPols = append(placementPols, &vimtypes.VmToVmGroupsAntiAffinity{
+			AntiAffinedVmGroupTags: requiredHostLabels,
+			PolicyStrictness:       string(vimtypes.VmPlacementPolicyVmPlacementPolicyStrictnessRequiredDuringPlacementPreferredDuringExecution),
+			PolicyTopology:         string(vimtypes.VmPlacementPolicyVmPlacementPolicyTopologyHost),
+		})
+	}
+
+	// Process preferred anti-affinity terms associated with zone topology.
+	preferredZoneLabels := extractTermLabelsByTopology(
+		vmCtx,
+		antiAffinity.PreferredDuringSchedulingPreferredDuringExecution,
+		corev1.LabelTopologyZone,
+	)
+	if len(preferredZoneLabels) > 0 {
+		placementPols = append(placementPols, &vimtypes.VmToVmGroupsAntiAffinity{
+			AntiAffinedVmGroupTags: preferredZoneLabels,
+			PolicyStrictness:       string(vimtypes.VmPlacementPolicyVmPlacementPolicyStrictnessPreferredDuringPlacementPreferredDuringExecution),
+			PolicyTopology:         string(vimtypes.VmPlacementPolicyVmPlacementPolicyTopologyVSphereZone),
+		})
+	}
+
+	// Process preferred anti-affinity terms associated with host topology.
+	preferredHostLabels := extractTermLabelsByTopology(
+		vmCtx,
+		antiAffinity.PreferredDuringSchedulingPreferredDuringExecution,
+		corev1.LabelHostname,
+	)
+	if len(preferredHostLabels) > 0 {
+		placementPols = append(placementPols, &vimtypes.VmToVmGroupsAntiAffinity{
+			AntiAffinedVmGroupTags: preferredHostLabels,
+			PolicyStrictness:       string(vimtypes.VmPlacementPolicyVmPlacementPolicyStrictnessPreferredDuringPlacementPreferredDuringExecution),
+			PolicyTopology:         string(vimtypes.VmPlacementPolicyVmPlacementPolicyTopologyHost),
+		})
+	}
+
+	return placementPols
+}
+
+// extractTermLabelsByTopology returns a list of labels extracted from VM's
+// affinity/anti-affinity terms with the specified topology key.
+func extractTermLabelsByTopology(
+	vmCtx pkgctx.VirtualMachineContext,
+	terms []vmopv1.VMAffinityTerm,
+	topologyKey string) []string {
+
+	var labels []string
+
+	for _, term := range terms {
+		if term.TopologyKey != topologyKey {
```

Comment:
I guess this is okay and cleaner, but a little inefficient.  There is a possibility of quadratic complexity because for each label type, we are going over all the affinity terms.  We could simply crawl these and create a map once for easy lookup. 
Not a blocker since what you have reads cleaner, and I don't anticipate that many affinity policies. But please model it on my suggestion once and see how it looks.

Reply:
> There is a possibility of quadratic complexity because for each label type, we are going over all the affinity terms.

This is not the case anymore since we've dropped processing host type label. The current code loops each following term list exactly once to get only the zone type labels:
- affinity.RequiredDuringSchedulingPreferredDuringExecution
- affinity.PreferredDuringSchedulingPreferredDuringExecution
- antiAffinity.RequiredDuringSchedulingPreferredDuringExecution
- antiAffinity.PreferredDuringSchedulingPreferredDuringExecution

So the complexity would be O(n) where n is the total number of terms defined in the fields above.